File: packages/core-js/src/detect.ts
import type { DetectResult } from './types.js';

const EXTENSIONS = ['.jsonmd', '.json.md', '.md'];

/**
 * Detect if a request should be converted to Markdown.
 * Returns the clean URL (extension stripped, format param removed) and whether to convert.
 */
export function detectRequest(
  url: string,
  headers?: Record<string, string | string[] | undefined>,
): DetectResult {
  const headerGet = (name: string): string | undefined => {
    if (!headers) return undefined;
    const val = headers[name] ?? headers[name.toLowerCase()];
    return Array.isArray(val) ? val[0] : val;
  };

  // Parse URL
  let pathname: string;
  let search: string;
  try {
    // Handle both full URLs and path-only
    if (url.startsWith('http')) {
      const parsed = new URL(url);
      pathname = parsed.pathname;
      search = parsed.search;
    } else {
      const qIdx = url.indexOf('?');
      pathname = qIdx >= 0 ? url.slice(0, qIdx) : url;
      search = qIdx >= 0 ? url.slice(qIdx) : '';
    }
  } catch {
    return { shouldConvert: false, cleanUrl: url };
  }

  // 1. Check URL extension
  for (const ext of EXTENSIONS) {
    if (pathname.endsWith(ext)) {
      const cleanPath = pathname.slice(0, -ext.length);
      return {
        shouldConvert: true,
        cleanUrl: cleanPath + search,
        ...parseModeFromSearch(search),
      };
    }
  }

  // 2. Check Accept header
  const accept = headerGet('accept');
  if (accept && accept.includes('text/markdown')) {
    return {
      shouldConvert: true,
      cleanUrl: url,
      ...parseModeFromSearch(search),
    };
  }

  // 3. Check query parameter
  if (search) {
    const params = new URLSearchParams(search);
    const format = params.get('format');
    if (format === 'md' || format === 'markdown') {
      params.delete('format');
      const mode = params.get('mode') as 'table' | 'signatures' | null;
      if (mode) params.delete('mode');
      const remaining = params.toString();
      const cleanUrl = pathname + (remaining ? '?' + remaining : '');
      return {
        shouldConvert: true,
        cleanUrl,
        ...(mode ? { mode } : {}),
      };
    }
  }

  // 4. Check X-Format header
  const xFormat = headerGet('x-format');
  if (xFormat && xFormat.toLowerCase() === 'markdown') {
    return {
      shouldConvert: true,
      cleanUrl: url,
      ...parseModeFromSearch(search),
    };
  }

  return { shouldConvert: false, cleanUrl: url };
}

function parseModeFromSearch(search: string): { mode?: 'table' | 'signatures' } {
  if (!search) return {};
  const params = new URLSearchParams(search);
  const mode = params.get('mode');
  if (mode === 'signatures' || mode === 'table') return { mode };
  return {};
}

File: packages/core-python/jsonmd/detect.py
"""Request detection for jsonmd conversion triggers."""

from __future__ import annotations

from dataclasses import dataclass, field
from urllib.parse import urlparse, parse_qs, urlencode


EXTENSIONS = [".jsonmd", ".json.md", ".md"]


@dataclass
class DetectResult:
    should_convert: bool = False
    clean_url: str = ""
    mode: str | None = None


def detect_request(url: str, headers: dict[str, str] | None = None) -> DetectResult:
    """Detect if a request should be converted to Markdown."""
    headers = headers or {}

    def header_get(name: str) -> str | None:
        return headers.get(name) or headers.get(name.lower())

    # Parse URL
    if "?" in url:
        path, query_string = url.split("?", 1)
    else:
        path, query_string = url, ""

    # 1. Check URL extension
    for ext in EXTENSIONS:
        if path.endswith(ext):
            clean_path = path[: -len(ext)]
            clean_url = clean_path + ("?" + query_string if query_string else "")
            mode = _parse_mode(query_string)
            return DetectResult(should_convert=True, clean_url=clean_url, mode=mode)

    # 2. Check Accept header
    accept = header_get("accept")
    if accept and "text/markdown" in accept:
        mode = _parse_mode(query_string)
        return DetectResult(should_convert=True, clean_url=url, mode=mode)

    # 3. Check query parameter
    if query_string:
        params = parse_qs(query_string, keep_blank_values=True)
        fmt = params.get("format", [None])[0]
        if fmt in ("md", "markdown"):
            new_params = {k: v for k, v in params.items() if k != "format"}
            mode_val = new_params.pop("mode", [None])[0] if "mode" in new_params else None
            remaining = urlencode(new_params, doseq=True)
            clean_url = path + ("?" + remaining if remaining else "")
            return DetectResult(should_convert=True, clean_url=clean_url, mode=mode_val)

    # 4. Check X-Format header
    x_format = header_get("x-format")
    if x_format and x_format.lower() == "markdown":
        mode = _parse_mode(query_string)
        return DetectResult(should_convert=True, clean_url=url, mode=mode)

    return DetectResult(should_convert=False, clean_url=url)


def _parse_mode(query_string: str) -> str | None:
    if not query_string:
        return None
    params = parse_qs(query_string)
    mode = params.get("mode", [None])[0]
    if mode in ("table", "signatures"):
        return mode
    return None
